### DISCUSSION TITLE
BaseModel: expose __init__ method arguments and class docstring like dataclasses

### DISCUSSION THREAD

**Discussion by tweakimp** [pydantic/pydantic#2464] [Feature Request]
I would like to see the BaseModel parent class to expose the `__init__` method header and the docstring like [dataclasses ](https://docs.python.org/3/library/dataclasses.html)do.
This feature reminds me of [functools.wraps](https://docs.python.org/3/library/functools.html#functools.wraps), which saves the name and the docstring of decorated functions.

Right now, the docstring of the `Inherited` class is missing when `BaseModel` is inherited and the `__init__` method is not exposed when writing the initialization of an instance.

Open the following code in your IDE (Python 3.7 or higher):
```python
from pydantic import BaseModel
from dataclasses import dataclass


@dataclass
class Dataclass:
    """
    Docstring.
    """
    string: str
    integer: int


class Inherited(BaseModel):
    """
    Docstring.
    """
    string: str
    integer: int


Dataclass()
Inherited()
```

In VSCode it looks like this:
Dataclass completion hint shows `__init__` arguments and types
![image](https://user-images.githubusercontent.com/20489521/109847729-d3a50500-7c4f-11eb-95a0-c4c6d647513d.png)
Dataclass hover shows arguments, types and docstring
![image](https://user-images.githubusercontent.com/20489521/109847947-0949ee00-7c50-11eb-87df-eaf93c76f767.png)
Inherited completion hint only says `**data`
![image](https://user-images.githubusercontent.com/20489521/109847980-11a22900-7c50-11eb-8c6d-356c1b7c4cc7.png)
Inherited hover also inherits BaseModel docstring and overwrites class docstring
![image](https://user-images.githubusercontent.com/20489521/109847993-1535b000-7c50-11eb-9156-a874ec1b9d99.png)

**Comment by michaeloliverx** [comment]
AFAIK this isn't possible because all this data is populated by metaclass magic. For pycharm there is a [plugin](https://pydantic-docs.helpmanual.io/pycharm_plugin/). I wonder if we could get special support in **MyPy**/**Pylance** for pydantic classes/dataclasses. **MyPy** has [special support for the `attrs` package](https://mypy.readthedocs.io/en/stable/additional_features.html#the-attrs-package) so it may be possible ? I know there is a `mypy` plugin so we probably only need support to be added to [Pylance](https://github.com/microsoft/pylance-release) to help with the VSCode auto-complete.

**Reply by michaeloliverx** [reply]
We can [convert a standard library dataclass into a pydantic dataclass](https://pydantic-docs.helpmanual.io/usage/dataclasses/#stdlib-dataclasses-and-pydantic-dataclasses):

```python
import dataclasses

import pydantic


@pydantic.dataclasses.dataclass
@dataclasses.dataclass
class ErrorResponse:
    """
    This is my lovely docstring.
    """

    code: str
    message: str


ErrorResponse()
```

This helps with the VSCode docstring a little:
<img width="536" alt="image" src="https://user-images.githubusercontent.com/55017335/110149835-68396f80-7dd6-11eb-8719-f3ddb757bb04.png">


but it doesn't support the arguments:
<img width="497" alt="image" src="https://user-images.githubusercontent.com/55017335/110149757-52c44580-7dd6-11eb-80c2-c438827072f9.png">
